net/stunserver/stunserver.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package stunserver implements a STUN server. The package publishes a number of stats
+// to expvar under the top level label "stun". Logs are sent to the standard log package.
+package stunserver
+
+import (
+	"context"
+	"errors"
+	"expvar"
+	"io"
+	"log"
+	"net"
+	"net/netip"
+	"time"
+
+	"tailscale.com/metrics"
+	"tailscale.com/net/stun"
+)
+
+var (
+	stats           = new(metrics.Set)
+	stunDisposition = &metrics.LabelMap{Label: "disposition"}
+	stunAddrFamily  = &metrics.LabelMap{Label: "family"}
+	stunReadError   = stunDisposition.Get("read_error")
+	stunNotSTUN     = stunDisposition.Get("not_stun")
+	stunWriteError  = stunDisposition.Get("write_error")
+	stunSuccess     = stunDisposition.Get("success")
+
+	stunIPv4 = stunAddrFamily.Get("ipv4")
+	stunIPv6 = stunAddrFamily.Get("ipv6")
+)
+
+func init() {
+	stats.Set("counter_requests", stunDisposition)
+	stats.Set("counter_addrfamily", stunAddrFamily)
+	expvar.Publish("stun", stats)
+}
+
+type STUNServer struct {
+	ctx context.Context // ctx signals service shutdown
+	pc  *net.UDPConn    // pc is the UDP listener
+}
+
+// New creates a new STUN server. The server is shutdown when ctx is done.
+func New(ctx context.Context) *STUNServer {
+	return &STUNServer{ctx: ctx}
+}
+
+// Listen binds the listen socket for the server at listenAddr.
+func (s *STUNServer) Listen(listenAddr string) error {
+	pc, err := net.ListenPacket("udp", listenAddr)
+	if err != nil {
+		return err
+	}
+	s.pc = pc.(*net.UDPConn)
```

Comment:
You could use `net.ListenUDP` instead to avoid the type assertion? Though I guess parsing the `listenAddr` into a `*UDPAddr` is annoying, so feel free to ignore.

Reply:
yeah, this part of the net package just has some razors, it's not really worth trying to optimize